common/rpc/interceptor/sdk_version.go

```diff
@@ -0,0 +1,114 @@
+// The MIT License
+//
+// Copyright (c) 2022 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package interceptor
+
+import (
+	"context"
+	"sync"
+
+	"go.temporal.io/server/common/headers"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/version/check"
+	"google.golang.org/grpc"
+)
+
+type sdkNameVersion struct {
+	name    string
+	version string
+}
+
+type SDKVersionInterceptor struct {
+	sdkInfoSet           map[sdkNameVersion]bool
+	lock                 sync.RWMutex
+	infoSetSizeCapGetter func() int
+	metricsClient        metrics.Client
+}
+
+var _ grpc.UnaryServerInterceptor = (*TelemetryInterceptor)(nil).Intercept
+
+func NewSDKVersionInterceptor(infoSetSizeCapGetter func() int, metricsClient metrics.Client) *SDKVersionInterceptor {
+	return &SDKVersionInterceptor{
+		sdkInfoSet:           make(map[sdkNameVersion]bool),
+		lock:                 sync.RWMutex{},
+		infoSetSizeCapGetter: infoSetSizeCapGetter,
```

Comment:
What is the purpose of this getter? Can't you just have a fixed size? If the size doesn't change at runtime, then just accept a normal int here.

Reply:
Can you just pass in the result of the getter during construction instead of storing the getter?

Reply:
But then it wouldn't make sense to call this dynamic config

Reply:
Not sure such a config needs to be dynamic (such an "extreme max" value really is better as just a const IMO so as not to add noise to the pieces that actually deserve to be configurable). But :shrug: it matters little.